src/Microsoft.AspNetCore.SignalR.Client/RpcConnection.cs

```diff
@@ -12,33 +13,47 @@
 
 namespace Microsoft.AspNetCore.SignalR.Client
 {
-    public class RpcConnection : IDisposable
+    public class RpcConnection : IDisposable, IInvocationBinder
     {
-        private readonly IInvocationAdapter _adapter;
-        private readonly Connection _connection;
-        private readonly Stream _stream;
         private readonly Task _reader;
+        private readonly Stream _stream;
+        private readonly ILogger _logger;
+        private readonly Connection _connection;
+        private readonly IInvocationAdapter _adapter;
 
         private readonly CancellationTokenSource _readerCts = new CancellationTokenSource();
         private readonly ConcurrentDictionary<string, InvocationRequest> _pendingCalls = new ConcurrentDictionary<string, InvocationRequest>();
+        private readonly ConcurrentDictionary<string, InvocationHandler> _handlers = new ConcurrentDictionary<string, InvocationHandler>();
 
         private int _nextId = 0;
 
-        private RpcConnection(Connection connection, IInvocationAdapter adapter)
+        private RpcConnection(Connection connection, IInvocationAdapter adapter, ILogger logger)
         {
             _connection = connection;
             _stream = connection.GetStream();
             _adapter = adapter;
+            _logger = logger;
+
             _reader = ReceiveMessages(_readerCts.Token);
         }
 
+        // TODO: Client return values/tasks?
+        // TODO: Overloads that use type parameters (like On<T1>, On<T1, T2>, etc.)
```

Comment:
I commented on this above actually... It was implemented as extensions before: https://github.com/SignalR/SignalR/blob/dev/src/Microsoft.AspNet.SignalR.Client/HubProxyExtensions.cs#L122